syntax = "proto3";

package RawTrace;

message Syscall {
	uint64 num = 1;
	uint32 id = 2;
	string name = 3;
	uint32 nr_args = 4;
	repeated string arg_names = 5;
	repeated uint64 arg_values = 6;
}

message Mem {
	uint64 addr = 1;
	uint64 value = 2;
	uint64 ip = 3;
	uint32 size = 4;
	uint32 pid = 5;
	bool is_write = 6;
}

message Ins {
	uint64 num = 1;
	uint64 pc = 2;
	uint64 rax = 3;
	uint64 rcx = 4;
	uint64 rdx = 5;
	uint64 rbx = 6;
	uint64 rsp = 7;
	uint64 rbp = 8;
	uint64 rsi = 9;
	uint64 rdi = 10;
	uint64 r8 = 11;
	uint64 r9 = 12;
	uint64 r10 = 13;
	uint64 r11 = 14;
	uint64 r12 = 15;
	uint64 r13 = 16;
	uint64 r14 = 17;
	uint64 r15 = 18;
	uint64 cc_op = 19;
	uint64 cc_dep1 = 20;
	uint64 cc_dep2 = 21;
	uint64 cc_ndep = 22;
	uint64 gs = 23;
	uint64 r_184 = 24;
	uint64 r_192 = 25;
	uint64 r_200 = 26;
	uint64 pid = 27;
}

enum DataType {
	EXPORT_INVALID = 0;
	EXPORT_KMEM_CACHE_OBJECT_SIZE = 1;
	EXPORT_KMEM_CACHE_OBJECT_PTR = 2;
}

message Data {
	DataType type = 1;
	uint64 value = 2;
}

enum SchedType {
	SCHED_INVALID = 0;
	SCHED_THREAD_CREATE = 1;
	SCHED_THREAD_OUT = 2;
	SCHED_THREAD_IN = 3;
	SCHED_THREAD_EXIT = 4;
	SCHED_SOFTIRQ_CREATE = 5;
	SCHED_SOFTIRQ_OUT = 6;
	SCHED_SOFTIRQ_IN = 7;
	SCHED_WORK_CREATE = 8;
	SCHED_WORK_OUT = 9;
	SCHED_WORK_IN = 10;
}

message Sched {
	SchedType type = 1;
	uint32 pid = 2;
	uint64 target = 3;
}

message Record {
	oneof item {
		Ins ins = 1;
		Mem mem = 2;
		Data data = 3;
		Syscall syscall = 4;
		Sched sched = 5;
	}
}

message Trace {
	repeated Record records = 1;
}

